darwin/importer/importer.py

```diff
@@ -205,20 +210,41 @@ def import_annotations(
     else:
         remote_classes = build_main_annotations_lookup_table(team_classes)
 
-    # Need to re parse the files since we didn't save the annotations in memory
-    for local_path in set(local_file.path for local_file in local_files):
-        parsed_files = importer(local_path)
-        if type(parsed_files) is not list:
-            parsed_files = [parsed_files]
-        # remove files missing on the server
-        missing_files = [missing_file.full_path for missing_file in local_files_missing_remotely]
-        parsed_files = [parsed_file for parsed_file in parsed_files if parsed_file.full_path not in missing_files]
-        for parsed_file in track(parsed_files):
+    with ThreadPoolExecutor(max_workers=max_workers) as executor:
+
+        def get_parsed_files(local_path):
+            parsed_files = importer(local_path)
+            if type(parsed_files) is not list:
+                parsed_files = [parsed_files]
+            # remove files missing on the server
+            missing_files = [missing_file.full_path for missing_file in local_files_missing_remotely]
+            parsed_files = [parsed_file for parsed_file in parsed_files if parsed_file.full_path not in missing_files]
+            return parsed_files
+
+        def import_annotation(parsed_file):
             image_id = remote_files[parsed_file.full_path]
             _import_annotations(
-                dataset.client, image_id, remote_classes, attributes, parsed_file.annotations, dataset, append
+                dataset.client,
+                image_id,
+                remote_classes,
+                attributes,
+                parsed_file.annotations,
+                dataset,
+                append,
             )
 
+        print(f"Uploading annotations with {max_workers} workers")
+        data = set(local_file.path for local_file in local_files)
+        # let's get the files we need to process
+        stage = executor.map(get_parsed_files, data)
+        out = list(stage)
+        parsed_files = chain.from_iterable(out)
+        # we have to get the total size, so let's convert parsed_files to list
+        parsed_files = list(parsed_files)
+        # process the file
+        stage = track(executor.map(import_annotation, parsed_files), total=len(parsed_files))
+        list(stage)
```

Comment:
It's not immediately easy for me to understand this code. That's probably a sign it needs a few comments to document the steps we're taking?

Reply:
updated the comments